backplane/datetimecalc
# enhance parsing functionality, migrate to uv

This PR modernizes the build system by migrating to the uv package manager and introduces significant enhancements to timedelta handling, timezone operations, and CLI functionality. 

---
🚀 Build System & Dependency Management

- Migrate to uv package manager with uv.lock for reproducible builds
- Move dependencies to pyproject.toml (removed requirements.txt and requirements-dev.txt)
- Set minimum Python version to 3.12
- Add dependency-groups for dev dependencies (pytest>=8.4.2)
- Update GitHub Actions workflows to latest versions:
  - actions/checkout@v5
  - actions/setup-python@v6
  - actions/upload-artifact@v5
- Update pre-commit configuration:
  - Use uv run pytest instead of direct pytest
  - Disable python-safety-dependencies-check (no uv support yet)
  - Update pre-commit hook versions

---
✨ Core Functionality Enhancements

Timedelta Improvements

- Refactored TDC.extract_components() with iterative algorithm
  - Proper handling of negative timedeltas
  - More accurate component extraction using total_seconds()
  - Sign preservation across all time units

New Parsing Functions

- parse_repr_str() - Parse Python repr() format strings back into datetime/timedelta objects
- spliton() - Utility for splitting sequences based on predicates (supports lt, le, eq, ne, ge, gt, is_, is_not, contains, isin, notin)
- offset_timezone() - Parse UTC offset strings in +05:30 and -08:00 formats

Timezone Enhancements

- Timezone comparison operations - Compare timezones by UTC offset using standard comparison operators
- Enhanced search_tz() - Now returns tuple of (tzinfo, filtered_input) allowing timezone extraction from strings
- Offset regex support - OFFSET_REGEX pattern for matching timezone offsets
- Improved fallback logic - Better handling of local timezone abbreviations and DST variants

---
🎯 CLI Improvements

- New --repr flag - Output Python repr() format instead of human-friendly strings
- Multi-word expression support - CLI now joins all arguments into single expression
- Better error messages - Comprehensive debug logging throughout parsing pipeline

---
🐛 Bug Fixes & Code Quality

- Fix is_timezone() docstring - Was incorrectly copied from is_timedelta()
- Add boolean formatting - format_temporal_object() now handles boolean results from comparisons
- Improve parse accuracy - parse_datetime_str() now checks parse accuracy and rejects zero-accuracy results
- Enhanced timezone filtering - parse_datetime_str() uses filtered input after timezone extraction
- Standardize docstrings - Remove leading spaces after opening quotes for consistency
- Type hints improvements - Better type annotations throughout codebase
